src/view/petgraph.rs

```diff
@@ -38,27 +39,37 @@ impl petgraph::visit::NodeRef for NodeIndex {
     }
 }
 
+/// Implement petgraph traits for a portgraph type or wrapper.
+///
+/// Can be called simply as `impl_petgraph_traits!(MyGraph)` for non-generic
+/// graphs, or
+/// ```ignore
+/// impl_petgraph_traits!(MyGraph<T, G>, [T, G] where T: Bound, G: OtherBound);
+/// ```
+/// for generic types with bounds.
 macro_rules! impl_petgraph_traits {
-    ($graph:ident) => {
-        impl petgraph::visit::GraphBase for $graph {
+    ($graph:ty) => {impl_petgraph_traits!($graph, []);};
+    ($graph:ty, [$($args:tt)*]) => {impl_petgraph_traits!($graph, [$($args)*] where );};
```

Comment:
What is the name of the thing you're doing with `where` here, so I can look it up?

Reply:
It's a hacky [macros by example](https://doc.rust-lang.org/reference/macros-by-example.html) match expression. The where is just a literal to match on so the call sites look sightly less confusing.

The full expression is:
```
$graph:ty,      // Match a type expression (can include '<...>'), assign it to variable $graph
[$($args:tt)*]  // Match a list of token trees (i.e. any expression with a single root).
                //   We use it for declaring type variables and lifetimes
where           // Match an explicit "where" token,
                //   so we can separate the variables from what comes next more easily
$($where:tt)*   // Match a list of token trees (can be empty).
                //   We use it for the bounds.
```
and the smaller expressions are variants with less elements that call the bigger one with some default.

We could do more strict match expressions instead of the catch-all "token-tree" (`:tt`), but since this is an internal macro it's not that critical.